packages/components/src/avatar/avatar.tsx

```diff
@@ -1,97 +1,356 @@
+    }
+  | {
+      type: 'channel'
+      size: 80
+      emoji: string
+      outline?: Variants['outline']
+      backgroundColor?: ColorTokens
+      background?: ColorTokens
+    }
+  | {
+      type: 'community'
+      size: 80 | 32 | 24 | 20
+      name: string
+      src?: string
+      outline?: Variants['outline']
+      backgroundColor?: ColorTokens
+    }
+  | {
+      type: 'account'
+      size: 80 | 48 | 32 | 28 | 24 | 20
+      name: string
+      src?: string
+      outline?: Variants['outline']
+      backgroundColor?: ColorTokens
+    }
+  | {
+      type: 'icon'
+      size: 48 | 32 | 20
+      icon: React.ReactElement
+      outline?: Variants['outline']
+      backgroundColor?: ColorTokens
+      color?: ColorTokens
+    }
 
 type ImageLoadingStatus = 'idle' | 'loading' | 'loaded' | 'error'
 
-const Avatar = (props: Props) => {
-  const {
-    src,
-    size,
-    shape = 'circle',
-    outline = false,
-    indicator = 'none',
-  } = props
+// todo?: use tokens
+const paddingSizes: Record<
+  NonNullable<AvatarProps['size']>,
+  number | undefined
+> = {
+  '80': 4,
+  '56': 2,
+  '48': 2,
+  '32': 2,
+  '28': undefined,
+  '24': undefined,
+  '20': undefined,
+  '16': undefined,
+}
+
+const radiusSizes: Record<
+  NonNullable<AvatarProps['size']>,
+  number | undefined
+> = {
+  '80': 16,
+  '56': undefined,
+  '48': 12,
+  '32': 10,
+  '28': 8,
+  '24': 8,
+  '20': 6,
+  '16': undefined,
+}
+
+const emojiSizes: Record<AvatarProps['size'], TextProps['size'] | undefined> = {
+  '80': 32,
+  '56': undefined,
+  '48': undefined,
+  '32': 15,
+  '28': undefined,
+  '24': 13,
+  '20': 11,
+  '16': undefined,
+}
+
+const textSizes: Record<NonNullable<AvatarProps['size']>, TextProps['size']> = {
+  '80': 32,
+  '56': 27,
+  '48': 19,
+  '32': 15,
+  '28': 13,
+  '24': 13,
+  '20': 11,
+  '16': 11,
+}
+
+const iconSizes: Record<
+  NonNullable<AvatarProps['size']>,
+  React.ReactElement
+> = {
+  '80': <MembersIcon20 />,
+  '56': <></>,
+  '48': <MembersIcon20 />,
+  '32': <MembersIcon16 />,
+  '28': <MembersIcon16 />,
+  '24': <></>,
+  '20': <MembersIcon12 />,
+  '16': <></>,
+}
+
+function hasIdenticon(
+  props: AvatarProps
+): props is Extract<AvatarProps, { colorHash?: number[][] }> {
+  return (
+    (props as Extract<AvatarProps, { colorHash?: number[][] }>).colorHash !==
+    undefined
+  )
+}
+
+function hasLock(
+  props: AvatarProps
+): props is Extract<AvatarProps, { lock?: 'locked' | 'unlocked' }> {
+  return (
+    (props as Extract<AvatarProps, { lock?: 'locked' | 'unlocked' }>).lock !==
+    undefined
+  )
+}
+
+function hasImage(
+  props: AvatarProps
+): props is Extract<AvatarProps, { src?: string }> {
+  return (props as Extract<AvatarProps, { src?: string }>).src !== undefined
+}
+
+const Avatar = (props: AvatarProps) => {
+  const { type, size, backgroundColor = '$neutral-95', outline = false } = props
+
+  // todo?: conditional hook(s)
+
+  const colorHash = hasIdenticon(props) ? props.colorHash : null
```

Comment:
```suggestion
const colorHash = 'colorHash' in props ? props.colorHash : undefined
```

I usually write it like this which does not require a separate guard function.